cpp/include/cugraph/mtmg/instance_manager.hpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph/mtmg/handle.hpp>
+
+#include <nccl.h>
+
+#include <vector>
+
+namespace cugraph {
+namespace mtmg {
+
+/**
+ * @brief Manages a subset of the cluster for a set of graph computations
+ */
+class instance_manager_t {
+ public:
+  /**
+   * @brief Constructor
+   *
+   * @param handles   Vector of RAFT handles, one for each device on this node
+   */
+  instance_manager_t(std::vector<std::shared_ptr<raft::handle_t>>&& handles,
+                     std::vector<std::shared_ptr<ncclComm_t>>&& nccl_comms,
+                     std::vector<rmm::cuda_device_id>&& device_ids)
+    : thread_counter_{0}, raft_handle_{handles}, nccl_comms_{nccl_comms}, device_ids_{device_ids}
+  {
+  }
+
+  /**
+   * @brief Get handle
+   *
+   * The instance manager will construct a handle appropriate for the thread making
+   * the request.  Threads will be assigned to GPUs in a round-robin fashion to
+   * spread requesting threads around the GPU resources.
+   *
+   * This function will be CPU thread-safe.
+   *
+   * @return a handle for this thread.
+   */
+  handle_t get_handle()
+  {
+    int local_id = thread_counter_++;
```

Comment:
Sorry for nitpicking but `auto local_gpu_id = thread_counter_++ % raft_handle_.size()` might be cleaner? (avoid duplicating local_id % raft_handle_.size()?)